### PR TITLE
LibCrypto: Add Poly1305

### PR DESCRIPTION
**Pull request by stelar7**
Part 1/3 of adding ChaCha20-Poly1305 to LibTLS

### CODE REVIEW

**Comment by bgianfo** [comment]
We should probably add a fuzzer, for this, see: https://github.com/SerenityOS/serenity/blob/master/Meta/Lagom/Fuzzers/FuzzSHA512.cpp

**Comment by stelar7** [comment]
> We should probably add a fuzzer, for this, see: https://github.com/SerenityOS/serenity/blob/master/Meta/Lagom/Fuzzers/FuzzSHA512.cpp

Since it requires 32 bytes of initial state, should those somehow be taken from the fuzzer? Or just use a fixed input for that?

**Comment by IdanHo** [comment]
> Since it requires 32 bytes of initial state, should those somehow be taken from the fuzzer? Or just use a fixed input for that?

You can just return 0 when `size` is not at least 32. Their coverage-guided fuzzers will understand that pretty fast and start giving you more input. Then any data above 32 can go into `update`.

File: Meta/Lagom/Fuzzers/FuzzPoly1305.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCrypto/Authentication/Poly1305.h>
+#include <stddef.h>
+#include <stdint.h>
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
+{
+    if (size < 32)
```
Comment (alimpfard) [inline Meta/Lagom/Fuzzers/FuzzPoly1305.cpp:13]: Maybe allow it to test empty messages too?

File: Userland/Libraries/LibCrypto/Authentication/Poly1305.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    u32 s[4] {};
+    u64 a[8] {};
+    u8 blocks[17] {};
+    u8 block_count {};
+};
+
+class Poly1305 {
+
+public:
+    void init(ReadonlyBytes key);
```
[the remark this answers is not in the document]
Comment (stelar7) [reply Userland/Libraries/LibCrypto/Authentication/Poly1305.h]: Fixed

File: Userland/Libraries/LibCrypto/Authentication/Poly1305.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    u32 s[4] {};
+    u64 a[8] {};
+    u8 blocks[17] {};
+    u8 block_count {};
+};
+
+class Poly1305 {
+
+public:
+    Poly1305(ReadonlyBytes key);
```
Comment (alimpfard) [inline Userland/Libraries/LibCrypto/Authentication/Poly1305.h:24]: pls `explicit` this.

**Review by alimpfard — COMMENTED** [review]
Just one more nitpick :P

File: Userland/Libraries/LibCrypto/Authentication/Poly1305.h (hunk 3 above)
Comment (stelar7) [reply Userland/Libraries/LibCrypto/Authentication/Poly1305.h]: Changed

**Pull request merged**